include/utils/views/dfs.h

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2022 Ultimaker B.V.
+// CuraEngine is released under the terms of the AGPLv3 or higher
+
+#ifndef CURAENGINE_DFS_SORT_H
+#define CURAENGINE_DFS_SORT_H
+
+#include <range/v3/algorithm/contains.hpp>
+#include <range/v3/view/subrange.hpp>
+
+#include "utils/concepts/graph.h"
+
+namespace cura::actions
+{
+/* # dfs utility
+ *
+ * Walks through an arbitrary graph using depth-first-search and calls a custom callback at each visited node
+ * \param node the current visited node, should be convertable or have the same type as the nodes in dag and visited
+ * \param dag Directed Acyclic Graph as defined by \concept isGraph
+ * \param state a state that p
+ * \param handle_node Custom call back function called at each visited node. Arguments for the functions are the current node, and the state resulted from the parent node
+ * \param visited nodes as defined by concept \isSet; _note: visited will not be ordered if it is of type `unordered_###`
+ */
+
+template <typename Node, typename State>
+constexpr void dfs(
+    const Node& current_node,
+    const isGraph auto& dag,
+    const State& state,
+    std::function<State(const Node, const State)> handle_node,
+    isSet auto& visited)
+{
+    if (ranges::contains(visited, current_node))
```

Comment:
Please note that the `unordered_***.find()` is faster then the `ranges::contains()` maybe use an `if constexpr` to check against the actual type used and deviate if needed.

This function is also a good candidate for a benchmark

Reply:
https://github.com/Ultimaker/CuraEngine/pull/1761/commits/cb267ffbba9d5f073255989933d7e599b00a6426